ttnn/cpp/ttnn/tensor/xtensor/partition.cpp

```diff
@@ -0,0 +1,168 @@
+        indices[dim] = xt::range(start, end);
+
+        auto chunk_view = xt::strided_view(xtensor, indices);
+
+        // Construct xarray from the view
+        // This forces a copy of that slice into a new xarray
+        chunks.push_back(xt::xarray<T>(chunk_view));
+        start = end;
+    }
+
+    return chunks;
+}
+
+template <typename T>
+xt::xarray<T> concatenate(const std::vector<xt::xarray<T>>& v, int dim) {
+    constexpr size_t MAX_TUPLE_SIZE = 64;
+
+    if (v.empty()) {
+        return {};
+    } else if (v.size() == 1) {
+        return v.front();
+    } else if (v.size() > MAX_TUPLE_SIZE) {
+        TT_THROW(
+            "Number of tensors to concatenate exceeds the maximum supported size {}; got size: {}",
+            MAX_TUPLE_SIZE,
+            v.size());
+    } else {
+        constexpr auto table = create_array<T, MAX_TUPLE_SIZE>();
+        return (*table[v.size()])(v, dim);
+    }
+}
+
+template xt::xarray<double> concatenate(const std::vector<xt::xarray<double>>& v, int dim);
+template xt::xarray<float> concatenate(const std::vector<xt::xarray<float>>& v, int dim);
+template xt::xarray<uint32_t> concatenate(const std::vector<xt::xarray<uint32_t>>& v, int dim);
+template xt::xarray<int32_t> concatenate(const std::vector<xt::xarray<int32_t>>& v, int dim);
+
+// Adaptor APIs from xtensor to ttnn::Tensor.
+namespace adaptor {
+namespace {
+
+template <typename VectorType, DataType TensorType>
+Tensor concatenate_impl(const std::vector<Tensor>& tensors, int dim) {
+    std::vector<xt::xarray<VectorType>> xtensors;
+    for (const auto& tensor : tensors) {
+        xtensors.push_back(to_xtensor<VectorType>(tensor));
+    }
+    xt::xarray<VectorType> result = concatenate(xtensors, dim);
+    return from_xtensor<VectorType, TensorType>(result);
+}
+
+template <typename VectorType, DataType TensorType>
+std::vector<Tensor> chunk_impl(const Tensor& tensor, int num_chunks, int dim) {
+    xt::xarray<VectorType> xtensor = to_xtensor<VectorType>(tensor);
+    auto xtensor_chunks = chunk<VectorType>(xtensor, num_chunks, dim);
+
+    std::vector<Tensor> tensors;
+    tensors.reserve(xtensor_chunks.size());
+    for (const auto& c : xtensor_chunks) {
+        tensors.push_back(from_xtensor<VectorType, TensorType>(c));
+    }
+    return tensors;
+}
+
+}  // namespace
+}  // namespace adaptor
+
+std::vector<Tensor> chunk(const Tensor& tensor, int num_chunks, int dim) {
+    switch (tensor.dtype()) {
+        case DataType::BFLOAT16: return adaptor::chunk_impl<float, DataType::BFLOAT16>(tensor, num_chunks, dim);
+        case DataType::FLOAT32: return adaptor::chunk_impl<float, DataType::FLOAT32>(tensor, num_chunks, dim);
+        case DataType::INT32: return adaptor::chunk_impl<int32_t, DataType::INT32>(tensor, num_chunks, dim);
+        case DataType::UINT32: return adaptor::chunk_impl<uint32_t, DataType::UINT32>(tensor, num_chunks, dim);
+        default: TT_THROW("Unsupported data type: {}", tensor.dtype());
+    }
```

Comment:
There's some existing scaffolding infra for dispatch on runtime dtype in `ttnn/cpp/ttnn/tensor/tensor_impl_wrapper.hpp` that you might be able to leverage and avoid the boilerplate here.

Reply:
As we generalize to other types, it will be used - thanks!